File: ts/src/sort/mergeSort.ts
export const mergeSort = <T>(arr: T[]): T[] => {
  const arrCopy = arr.slice();

  return mergeSortHelp(arrCopy, 0, arrCopy.length - 1);
}

const mergeSortHelp = <T>(arr: T[], minIndex: number, maxIndex: number): T[] => {
  if (minIndex >= maxIndex) {
    return arr;
  }
  const mid = Math.floor((minIndex + maxIndex) / 2);


  const left = mergeSortHelp(arr.slice(0, mid + 1), 0, mid);
  const right = mergeSortHelp(arr.slice(mid, arr.length), mid + 1, arr.length - 1);

  return merge(left, right);
};

const merge = <T>(arrLeft: T[], arrRight: T[]): T[] => {
  const result: T[] = [];

  let i = 0; // result index
  let j = 0; // left index
  let k = 0; // right index

  while (i < arrLeft.length + arrRight.length) {
    if (arrLeft[j] >= arrRight[k]) {
      result.push(arrLeft[j]);
      j++;
    } else {
      result.push(arrRight[k]);
      k++;
    }
    i++;
  }


  return result;
}

File: dataStructure/Java/src/main/java/sort/mergeSort/MergeSort.java
package sort.mergeSort;

import sort.common.SortFunction;

/**
 * Created by lzhao on 7/25/17.
 */

public class MergeSort implements SortFunction {
    public int[] sort(int[] array) {
        return mergeSort(array, 0, array.length - 1);
    }

    private int[] mergeSort(int[] array, int low, int high) {
        if (low >= high) {
            return new int[]{ array[high] };
        }

        int middle = (low + high) / 2;
        int[] left = this.mergeSort(array, low, middle);
        int[] right = this.mergeSort(array, middle + 1, high);

        return merge(left, right);
    }

    private int[] merge(int[] left, int[] right) {
        int[] result = new int[left.length + right.length];

        // left array index
        int i = 0;
        // right array index
        int j = 0;
        // result index
        int k = 0;

        while (true) {
            if (left[i] < right[j]) {
                result[k] = left[i];
                i++;
            } else {
                result[k] = right[j];
                j++;
            }
            k++;

            if (i >= left.length || j >= right.length) {
                break;
            }
        }

        while (i < left.length) {
            result[k] = left[i];
            i++;
            k++;
        }

        while (j < right.length) {
            result[k] = right[j];
            j++;
            k++;
        }

        return result;
    }
}
